ipc/provider/src/checkpoint.rs

```diff
@@ -0,0 +1,143 @@
+// Copyright 2022-2023 Protocol Labs
+// SPDX-License-Identifier: MIT
+//! Bottom up checkpoint manager
+
+use crate::config::Subnet;
+use crate::manager::{BottomUpCheckpointRelayer, EthSubnetManager};
+use anyhow::{anyhow, Result};
+use fvm_shared::address::Address;
+use fvm_shared::clock::ChainEpoch;
+use ipc_identity::{EthKeyAddress, PersistentKeyStore};
+use std::fmt::{Display, Formatter};
+use std::sync::{Arc, RwLock};
+use std::time::Duration;
+
+/// Tracks the metadata required for both top down and bottom up checkpoint submissions, such as
+/// parent/child subnet and checkpoint period.
+pub struct CheckpointMetadata {
+    parent: Subnet,
+    child: Subnet,
+    period: ChainEpoch,
+}
+
+pub struct BottomUpCheckpointManager<P, C> {
+    metadata: CheckpointMetadata,
+    parent_handler: P,
+    child_handler: C,
+}
+
+impl<P: BottomUpCheckpointRelayer, C: BottomUpCheckpointRelayer> BottomUpCheckpointManager<P, C> {
+    pub async fn new(
+        parent: Subnet,
+        child: Subnet,
+        parent_handler: P,
+        child_handler: C,
+    ) -> Result<Self> {
+        let period = parent_handler
+            .checkpoint_period(&child.id)
+            .await
+            .map_err(|e| anyhow!("cannot get bottom up checkpoint period: {e}"))?;
+        Ok(Self {
+            metadata: CheckpointMetadata {
+                parent,
+                child,
+                period,
+            },
+            parent_handler,
+            child_handler,
+        })
+    }
+}
+
+impl BottomUpCheckpointManager<EthSubnetManager, EthSubnetManager> {
+    pub async fn new_evm_manager(
+        parent: Subnet,
+        child: Subnet,
+        keystore: Arc<RwLock<PersistentKeyStore<EthKeyAddress>>>,
+    ) -> Result<Self> {
+        let parent_handler =
+            EthSubnetManager::from_subnet_with_wallet_store(&parent, keystore.clone())?;
+        let child_handler = EthSubnetManager::from_subnet_with_wallet_store(&child, keystore)?;
+        Self::new(parent, child, parent_handler, child_handler).await
+    }
+}
+
+impl<P: BottomUpCheckpointRelayer, C: BottomUpCheckpointRelayer> Display
+    for BottomUpCheckpointManager<P, C>
+{
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "bottom-up, parent: {:}, child: {:}",
+            self.metadata.parent.id, self.metadata.child.id
+        )
+    }
+}
+
+impl<
+        P: BottomUpCheckpointRelayer + Send + Sync + 'static,
```

Comment:
As mentioned above, I would try to have a single generic abstraction for the parent-child relay